_posts/2017-09-12-repeatable-serverless-applications.markdown

```diff
@@ -0,0 +1,108 @@
+At a recent Mechanical Rock mob programming session we decided to tackle the topic of, "How can we turn off any superfluous and idle ec2 instances" that we may have forgotten to stop/terminate – because servers are expensive right? We decided to call the new project, Instance Reaper. Source code available <a href="https://github.com/MechanicalRock/instance-reaper">here</a>. 
+
+![serverless benefits]({{ site.url }}/img/about_serverless_architecture.jpg)
+>Whilst a serverless architecture does alleviate a lot of these mundane tasks, it doesn't necessarily remove the need for them completely.
+- Serverless **is** scalable, fault tolerant and simplifies many of the operational tasks, associated with traditional architectures.
+
+
+The preferred approach for any new software project here at Mechanical Rock, is to tackle it with a test-first mindset and a serverless implementation. The difficulty comes in, with the infancy of the whole "serverless" paradigm. As a DevOps consultancy, ideally we would like to practice a Continuous Delivery model on all of our software projects. In order to achieve this, we need to be able test our application locally, on a pipeline and have it run bug-free in the cloud. The good news is, there are tools out there, that can help us achieve this.
+
+![serverless benefits]({{ site.url }}/img/localstack.png)
+
+Meet, LocalStack. LocalStack, was created by the good people at Atlassian and provides you with "A fully functional local AWS cloud stack". This means, that with a few cheeky bash scripts and some Docker magic, you can automate the provisioning of almost all the necessary resources for your cloud-based application. Sounds too easy? Well you're absolutely right! With a docker-compose file, you can start up the LocalStack service in conjunction with your dev-env in your Dockerfile. 
+
+```yml
+<!-- docker-compose.yml -->
+# add the folling service
+localstack:
+    image: localstack/localstack
+    ports:
+      - "4569-4581:4569-4581"
+      - "8080:8080"
+```
+
+In your dockerfile, you can reference a bash script which provisions your AWS resources so that when the docker services are up and running, your resources will be available on the relevant service port referenced on the <a href="https://github.com/localstack/localstack">LocalStack GitHub page</a>. If you would like visibility of the resources available on your 'stack', just hit up http://localhost:8080, in your browser. 
+
+```bash
+if [ $ENV = "local" ]; then
+    aws --endpoint-url=http://localhost:4572 s3 mb s3://instance-reaper-$COMPANY_NAME-logging
+else
+    aws s3 mb s3://instance-reaper-$COMPANY_NAME-logging --region $REGION
+fi
+```
+
+You can now test code locally, that can send and receive data from a dynamoDB table and leverage s3 buckets for file storage.
```

Comment:
DynamoDB is not included in the above snippet.

Reply:
You could achieve the same using mocks.

What makes Localstack more compelling over using mocks?

e.g. https://github.com/spulec/moto

Reply:
Containers with services always running versus, waiting for mocked services to be provisioned each time?

Reply:
You'll notice I mentioned that PynamoDB creates a table if the table doesn't already exist. Not sure if it's worth still including DyanmoDb in the shell snippet?